modules/http4s-emitter/src/main/scala/com.snowplowanalytics.snowplow.scalatracker.emitters.http4s/Http4sEmitter.scala

```diff
@@ -0,0 +1,146 @@
+        .compile
+        .drain
+    }
+
+  private def bufferEvents[F[_]](bufferConfig: BufferConfig): Pipe[F, Action, Request] =
+    _.mapAccumulate(Buffer(bufferConfig)) {
+      case (buffer, action) => buffer.handle(action)
+    }.collect {
+      case (_, Some(flushed)) => flushed
+    }
+
+  private def sendRequest[F[_]: Sync: Timer](client: Client[F],
+                                             collector: EndpointParams,
+                                             retryPolicy: RetryPolicy,
+                                             rng: Random,
+                                             callback: Option[Callback[F]]): Request => F[Unit] =
+    Sync[F].tailRecM(_) { request =>
+      attemptSend(client, collector, request)
+        .flatTap(invokeCallback(callback, collector, request, _))
+        .flatMap { result =>
+          if (result.isSuccess || request.isFailed(retryPolicy)) {
+            Sync[F].pure(Right(()))
+          } else {
+            for {
+              seed <- Sync[F].delay(rng.nextDouble())
+              _    <- Timer[F].sleep(RetryPolicy.getDelay(request.attempt, seed).millis)
+            } yield request.updateAttempt.asLeft
+          }
+        }
+    }
+
+  private def attemptSend[F[_]: MonadError[?[_], Throwable]: Timer](client: Client[F],
+                                                                    collector: EndpointParams,
+                                                                    request: Request): F[Result] =
+    Timer[F].clock.realTime(MILLISECONDS).flatMap { dtm =>
+      val httpRequest = request.updateStm(dtm) match {
+        case Request.Buffered(_, payload) =>
+          val body = Stream.emits(Payload.postPayload(payload).getBytes).covary[F]
+          HttpRequest[F](Method.POST, Uri.unsafeFromString(collector.getPostUri))
+            .withEntity(body)
+            .withContentType(`Content-Type`(MediaType.application.json))
+        case Request.Single(_, payload) =>
+          val uri =
+            Uri
+              .unsafeFromString(collector.getGetUri)
+              .setQueryParams {
+                payload.toMap.map { case (k, v) => k -> List(v) }
+              }
+          HttpRequest[F](Method.GET, uri)
+      }
+
+      client
+        .status(httpRequest)
+        .map { status =>
+          if (status.isSuccess) Result.Success(status.code)
+          else Result.Failure(status.code)
+        }
+        .handleError(Result.TrackerFailure(_))
+    }
+
+  private def invokeCallback[F[_]: MonadError[?[_], Throwable]](callback: Option[Callback[F]],
+                                                                collector: EndpointParams,
+                                                                request: Request,
+                                                                result: Result): F[Unit] =
+    callback match {
+      case Some(cb) =>
+        cb(collector, request, result).handleError(_ => ())
```

Comment:
I think we should force users to handle errors in callback, otherwise I can imagine a typical handling scenario where all unsent payloads are written to the disk and the disk runs out of space.

Reply:
I think what you are saying is that if `cb` raises an error, then we should ignore it here, and let it bubble up into the surrounding thread of execution.

This might not have the effect you originally think though.... the fiber that invokes the callback is completely isolated from the surrounding program's main fiber of execution.  So this fiber would die, but the program's main fiber would be unaware.  It is hard (not impossible but hard) to set things up in such a way that an exception in `ct` causes the program to exit.

The same is true of the AsyncEmitter in the id-emitters module.

I think I prefer to catch the exception, because it is easier to reason about the resulting behaviour.

Reply:
My reasoning was more about encouraging users to handle all errors rahter than closing an app (although that would be good as well). Because in case of `handleError(_ => ())` if something bad happens they won't even see it in logs, but ideally they should be encouraged to do something like sentry tracking.